How to search and order a phone number guide (#633)
Bandwidth/api-docs: site/docs/numbers/searchingNumbers.mdx

export const Highlight = ({children, color}) => (
    <span
        style={{
            backgroundColor: color,
            borderRadius: '2px',
            color: '#fff',
            padding: '0.2rem',
        }}>
      {children}
    </span>
);

> Ordering Phone Numbers for use with the network allows you to provision specific phone numbers from available inventory that have been discovered in a search.

Author: Seems like AJ already made a comment for this. Please check my updates for this.

In this guide we will show you how to make searching for available phone numbers and order them through Dashboard API.

## Search Numbers
You can find Bandwidth's available numbers using different criteria.
In certain cases where the **exact digits** are important (such as localVantiy, endsIn, npaNxxx) LCA should be disabled to filter out non-pattern matched phone numbers.
Bandwidth's available number searches with local calling enabled by default. The LCA flag searches nearby Ratecenters for phone numbers that are considered "local" to the parameters passed.

> To make a search for the available number, you must make a GET request to our [API available numbers](../../../../apis/numbers#tag/accounts/operation/GetAvailableTns) endpoint.

Reviewer: ```suggestion
To make a search for the available number, you must make a GET request to our [available numbers API](../../../../apis/numbers#tag/accounts/operation/GetAvailableTns) endpoint.
```

> This can be done through our [SDKs](/sdks), using tools like [Postman](https://github.com/Bandwidth/postman) or in the command line.

Reviewer: We dont link to the IRIS SDKs anywhere so id remove any reference to SDKs when talking about numbers apis

:::tip Searching Toll Free Numbers
The `tollFreeWildCardPattern` allows you to search toll free area codes, but forces the 3rd digit to be a `*`.
However, with the nature of Toll Free area codes - the 3rd digit will always be a repeat of the second, so to search for 833 numbers, set the parameter value as `83*` and you will only return 833 numbers.
For 800 numbers, search for `80*` and so on.
:::

> <Tabs

Reviewer: Is there a reason for replacing the different number search types with sdk examples? 

We dont really support the sdks and below is missing python and ruby

foreach (string number in result.TelephoneNumberList)
{
    Console.WriteLine(string.Format("Number: {0}", number));
}
```

</TabItem>
<TabItem value="nodejs">

Available Numbers Area Code

```nodejs
var res = await numbers.AvailableNumbers.listAsync({areaCode:"919", quantity:"2"});
console.log(res);
```

</TabItem>
<TabItem value="php">

```php
$account->availableNumbers([ "areaCode" => "818", "quantity" => "2" ]);
```

</TabItem>
</Tabs>

And endpoint `../availableNumbers` should return a response like:

> Response

```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<SearchResult>
    <ResultCount>2</ResultCount>
    <TelephoneNumberList>
        <TelephoneNumber>6672286710</TelephoneNumber>
        <TelephoneNumber>5756186700</TelephoneNumber>
    </TelephoneNumberList>
</SearchResult>
```

## Order Phone Numbers

To create an order, you can use the numbers found through the search.
If you want that Bandwidth will select phones for you according to your criteria and immediately create an order -
then you can look at the `orderType` parameter of the request schema in [API order numbers](../../../../apis/numbers#tag/accounts/operation/CreateOrder).

> In this example, we will show how to make an order for numbers that were found using the search.

Reviewer 2: ```suggestion
In this example, we will show how to create an order for numbers that were found using the search.
```

To order numbers, you must make a POST request to our [API order numbers](../../../../apis/numbers#tag/accounts/operation/CreateOrder) endpoint.

> This can be done through our [SDKs](/sdks), using tools like [Postman](https://github.com/Bandwidth/postman) or in the command line.

Reviewer 2: ```suggestion
This can be done through tools like [Postman](https://github.com/Bandwidth/postman) or in the command line.
```

        ]
    }
};
numbers.Order.create(order, callback);
```

</TabItem>

<TabItem value="php">

```php
$order = $account->orders()->create([
    "Name" => "Available Telephone Number order",
    "SiteId" => "461",
    "ExistingTelephoneNumberOrderType" => [
        "TelephoneNumberList" => [
            "TelephoneNumber" => [ "6672286710", "5756186700" ]
        ]
    ]
]);
```

</TabItem>

</Tabs>

And endpoint `../orders` should return a response like:

> Response

```xml
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<OrderResponse>
    <Order>
        <OrderCreateDate>2018-01-23T19:56:29.678Z</OrderCreateDate>
        <BackOrderRequested>false</BackOrderRequested>
        <id>47955555-67aa-4adb-8c0f-b6894e60c0dc</id>
        <ExistingTelephoneNumberOrderType>
            <TelephoneNumberList>
                <TelephoneNumber>6672286710</TelephoneNumber>
                <TelephoneNumber>5756186700</TelephoneNumber>
            </TelephoneNumberList>
        </ExistingTelephoneNumberOrderType>
        <PartialAllowed>true</PartialAllowed>
        <SiteId>461</SiteId>
    </Order>
    <OrderStatus>RECEIVED</OrderStatus>
</OrderResponse>
```

## Fetch Order Information

> Phone number ordering in the Bandwidth Dashboard is asynchronous when creating an "order". The orders are then processed and the order status is updated asynchronously.  Bandwidth recommends configuring your account with a subscription instead of polling the order resource for `<OrderStatus>`.

Reviewer 2: can you add a link to the subscriptions reference on the word "subscriptions"?

Order processing times can vary and are not guaranteed, so bandwidth does __not__ recommend setting a timeout on waiting for an order to show either `COMPLETE` or `FAILED` status, but instead relying on a webhook from an orders subscription.

> ## Poll for Order Status

Reviewer 2: Can we remove the poll section? I don't know that it makes sense to include since we don't recommend this method.

    <LastModifiedDate>2018-01-23T19:56:30.782Z</LastModifiedDate>
    <OrderCompleteDate>2018-01-23T19:56:30.782Z</OrderCompleteDate>
    <Order>
        <OrderCreateDate>2018-01-23T19:56:29.678Z</OrderCreateDate>
        <id>47955555-67aa-4adb-8c0f-b6894e60c0dc</id>
        <ExistingTelephoneNumberOrderType>
            <TelephoneNumberList>
                <TelephoneNumber>6672286710</TelephoneNumber>
                <TelephoneNumber>5756186700</TelephoneNumber>
            </TelephoneNumberList>
        </ExistingTelephoneNumberOrderType>
        <PartialAllowed>true</PartialAllowed>
        <SiteId>461</SiteId>
    </Order>
    <OrderStatus>COMPLETE</OrderStatus>
    <CompletedNumbers>
        <TelephoneNumber>
            <FullNumber>6672286710</FullNumber>
            <FullNumber>5756186700</FullNumber>
        </TelephoneNumber>
    </CompletedNumbers>
    <Summary>2 number ordered</Summary>
    <FailedQuantity>0</FailedQuantity>
</OrderResponse>
```

## Order Status Webhook

This is a webhook that bandwidth sends to _your_ server/email upon order completion/failure - there is no need to poll the resource if using this recommended method.

> To create a subscription for Order Status you must to send POST request to our [API create subscription](../../../../apis/numbers#tag/accounts/operation/CreateSubscriptions) endpoint.

Reviewer 2: ```suggestion
To create a subscription for Order Status you must send a POST request to our [API create subscription](../../../../apis/numbers#tag/accounts/operation/CreateSubscriptions) endpoint.
```

subscription.setOrderId("orderId");
EmailSubscription emailSubscription = new EmailSubscription();
emailSubscription.setEmail("test@test.com");
emailSubscription.setDigestRequested("NONE");
subscription.setEmailSubscription(emailSubscription);
Subscription createdSubscription = Subscription.create(client, subscription);
```

</TabItem>

<TabItem value="c#">

```c#

var subscription = await Subscription.Create(new Subscription() {
    OrderType = "orders",
    OrderId = "100",
    EmailSubscription = new EmailSubscription {
        Email = "test@test",
        DigestRequested = "NONE"
    }
};
```

</TabItem>

<TabItem value="nodejs">

```nodejs

var subscription = {
    orderType:"orders",
    callbackSubscription: {
        URL:"http://mycallbackurl.com",
        user:"userid",
        expiry: 12000
    }
};
numbers.Subscription.create(subscription, callback);
```

</TabItem>

<TabItem value="php">

```php
$subscription = $account->subscriptions()->create([
    "OrderType" => "portins",
    "OrderId" => "orderId",
    "EmailSubscription" => [
        "Email" => "test@test.com",
        "DigestRequested" => "DAILY"
    ]
]);
```

</TabItem>
</Tabs>

## Where to next?

> Now that you have learn how to search and order numbers, check out some of the available actions in the guides:

Reviewer 2: ```suggestion
Now that you have learned how to search and order numbers, check out some of the other available actions in our guides:
```
